### PR TITLE
Feature/issue 55 transaction edit form

### LINKED ISSUE #55: Ajouter/supprimer des lignes
# Gestion des lignes de template de budget - Epic: Templates de Budget

## **User Story 1: Ajouter et supprimer des lignes de template de budget**

**En tant qu'** utilisateur de Pulpe
**Je veux** pouvoir ajouter de nouvelles lignes et supprimer des lignes existantes dans un template de budget
**Afin de** adapter complètement le template à mes besoins spécifiques et maintenir uniquement les postes budgétaires pertinents

### Valeur métier

Cette fonctionnalité permet aux utilisateurs de :
- Personnaliser entièrement leurs templates de budget selon leur situation financière unique
- Éliminer les postes budgétaires non pertinents pour simplifier leur gestion
- Ajouter des postes spécifiques à leurs besoins (revenus exceptionnels, dépenses particulières)
- Maintenir des templates évolutifs qui s'adaptent aux changements de vie

#### Fonctionnalités principales

- [ ] **Ajout de nouvelle ligne** : L'utilisateur peut ajouter une nouvelle ligne de template avec les champs : nom, montant, type (INCOME/FIXED_EXPENSE/SAVINGS_CONTRIBUTION), récurrence (fixed/variable/one_off), et description
- [ ] **Suppression de ligne existante** : L'utilisateur peut supprimer n'importe quelle ligne existante du template
- [ ] **Confirmation de suppression** : Une boîte de dialogue de confirmation s'affiche avant la suppression d'une ligne
- [ ] **Minimum une ligne** : Le système empêche la suppression de la dernière ligne (minimum 1 ligne requis)

#### Sauvegarde et validation

- [ ] **Sauvegarde en lot (bulk)** : Toutes les modifications (ajouts + suppressions + éditions) sont sauvegardées en une seule opération transactionnelle
- [ ] **Validation côté client dynamique** : 
  - Les champs apparaissent en rouge uniquement s'ils ont été édités (touched) et sont invalides
  - Validation en temps réel : lignes non vides, montants positifs
  - Désactivation du bouton "Enregistrer" si le formulaire est invalide
- [ ] **Messages d'erreur explicites** : Si la sauvegarde échoue, un message d'erreur clair est affiché à l'utilisateur

#### Interface utilisateur

- [ ] **Interface temps réel** : L'ajout et la suppression de lignes sont reflétés immédiatement dans l'interface
- [ ] **État de chargement** : Les boutons sont désactivés et un indicateur de chargement s'affiche pendant la sauvegarde (déjà en place)
- [ ] **Calcul du total en cours** : La colonne "Total" affiche le montant cumulé en tenant compte du type de transaction

#### Scénario 1 : Ajout d'une nouvelle ligne
```
Étant donné que je suis sur la page d'édition d'un template de budget
Quand je clique sur "Ajouter une transaction"
Alors une nouvelle ligne vide est ajoutée au tableau
Et je peux saisir : nom, montant, type, et la description est pré-remplie
Et la nouvelle ligne est visible immédiatement sans rechargement
```

#### Scénario 2 : Suppression d'une ligne existante
```
Étant donné que j'ai un template avec plusieurs lignes
Quand je clique sur l'icône "supprimer" d'une ligne
Alors une boîte de dialogue de confirmation s'affiche
Et si je confirme, la ligne est supprimée immédiatement
Et le total est recalculé automatiquement
```

#### Scénario 3 : Protection contre suppression complète
```
Étant donné que j'ai un template avec une seule ligne
Quand je regarde l'icône "supprimer" de cette ligne
Alors le bouton est désactivé
Et un message ou tooltip indique qu'au moins une ligne est requise
```

#### Scénario 4 : Validation dynamique
```
Étant donné que j'ajoute une nouvelle ligne
Quand je saisis un montant négatif ou zéro
Et que je clique en dehors du champ (touched)
Alors le champ s'affiche en rouge avec un message d'erreur
Et le bouton "Enregistrer" est désactivé
```

#### Scénario 5 : Sauvegarde en lot
```
Étant donné que j'ai modifié plusieurs lignes, ajouté 2 nouvelles et supprimé 1 existante
Quand je clique sur "Enregistrer"
Alors toutes les modifications sont envoyées en une seule requête
Et soit toutes les modifications sont appliquées, soit aucune (atomicité)
Et un message de succès confirme la sauvegarde
```

#### Scénario 6 : Gestion d'erreur de sauvegarde
```
Étant donné que j'ai effectué des modifications
Quand la sauvegarde échoue (erreur réseau ou serveur)
Alors un message d'erreur explicite s'affiche
Et mes modifications restent visibles dans le formulaire
Et je peux réessayer la sauvegarde
```

### Règles métier applicables

Conformément au document SPECS.md :

- **RG-001 Template ↔︎ Budget Cohesion** : La modification d'un template peut impacter les budgets futurs créés à partir de ce template
- **Structure des templates** : Chaque ligne doit respecter la structure définie (revenus, dépenses fixes, contributions épargne)
- **Validation des montants** : Tous les montants doivent être positifs et réalistes (max 999'999 CHF)

#### Phase 1 : Corrections Frontend (30min)
- Supprimer la validation bloquante pour les ajouts/suppressions dans `EditTransactionsDialog`
- Implémenter la validation "touched" pour n'afficher les erreurs qu'après édition
- Ajouter une boîte de dialogue de confirmation pour les suppressions
- Gérer la protection contre la suppression de la dernière ligne

#### Phase 2 : Extension Backend (45min)
- Étendre l'API RPC pour supporter :
  - Création de nouvelles lignes (ID null dans la requête)
  - Suppression de lignes existantes (omission dans la liste)
- Ajouter validation côté serveur pour minimum 1 ligne
- Implémenter les tests unitaires pour les nouvelles fonctionnalités

#### Phase 3 : Tests E2E (30min)
- Tests complets des scénarios d'ajout/suppression
- Tests de validation et gestion d'erreurs
- Tests d'atomicité des opérations en lot

#### Structures de données concernées

```typescript
// Nouvelle ligne créée (frontend vers backend)
interface NewTemplateLine {
  id: null; // Indique une création
  name: string;
  amount: number;
  kind: TransactionKind;
  recurrence: TransactionRecurrence;
  description: string;
}

// Ligne supprimée : omise de la liste dans TemplateLinesBulkUpdate
```

#### Points d'attention

- **Atomicité** : L'opération de sauvegarde doit être transactionnelle (tout ou rien)
- **Performance** : Optimiser l'UI pour gérer de nombreuses lignes sans ralentissement
- **Accessibilité** : Assurer que les actions sont accessibles au clavier et aux lecteurs d'écran
- **Cohérence** : Maintenir la cohérence des calculs en temps réel lors des modifications

### Définition de "Terminé"

- [ ] L'utilisateur peut ajouter et supprimer des lignes de template
- [ ] Toutes les validations fonctionnent correctement
- [ ] La sauvegarde en lot est opérationnelle
- [ ] Les tests unitaires et E2E passent
- [ ] L'interface est responsive et accessible
- [ ] La documentation technique est mise à jour

---

# Instruction: Add and Delete Budget Template Lines

> Please follow this plan using proper rules.

## Goal
Implement the ability to add new lines and delete existing lines in budget templates, with bulk update operations and proper validation, while maintaining UI responsiveness and data integrity.

## Existing files
- `/frontend/projects/webapp/src/app/feature/budget-templates/details/components/edit-transactions-dialog.ts`
- `/backend-nest/src/modules/budget-template/budget-template.service.ts`
- `/backend-nest/src/modules/budget-template/budget-template.controller.ts`
- `/shared/schemas.ts` (Zod schemas)
- `/backend-nest/src/modules/budget-template/budget-template.mappers.ts`

## New file to create
- `/frontend/projects/webapp/src/app/feature/budget-templates/details/components/delete-line-confirmation.dialog.ts`

## Grouped tasks

- Remove blocking validation in `EditTransactionsDialog` lines 423-427 that prevents adding/removing lines
- Implement "touched" validation pattern to show errors only after user interaction with fields
- Add confirmation dialog component for line deletion using Angular Material Dialog
- Implement minimum 1 line protection by disabling delete button when only one line remains
- Update form validation logic to support dynamic add/remove operations
- Ensure proper signal reactivity when adding/removing lines from FormArray

- Modify `bulkUpdateTemplateLines` method to handle `id: null` as indicator for new line creation
- Implement line deletion by detecting omitted lines from the update array compared to existing lines
- Add server-side validation to enforce minimum 1 line requirement
- Ensure atomicity of all operations using Supabase transactions or RPC
- Update Zod schemas to allow optional `id` field in `TemplateLineUpdateWithId`
- Add proper error handling for validation failures

- Update frontend to send `id: null` for newly created lines in bulk update payload
- Modify delete operation to remove lines from update array instead of separate API calls
- Implement proper error handling for server validation failures
- Ensure UI state consistency during async operations
- Test complete workflow from add/edit/delete to successful save

- Add unit tests for bulk update logic handling creation and deletion
- Implement E2E tests covering add/remove scenarios with validation
- Test edge cases including minimum line validation and invalid data
- Verify atomicity ensuring all operations succeed or fail together
- Test UI responsiveness during loading states and error handling

## Validation checkpoints
- All add/remove operations work without validation errors
- Minimum 1 line protection prevents accidental deletion of last line
- Bulk update successfully creates, updates, and deletes lines atomically
- Proper error messages display for validation failures
- UI remains responsive with proper loading states
- E2E tests pass for complete user workflows

## Official Documentation References
- [Angular Reactive Forms](https://angular.dev/guide/forms/reactive-forms)
- [Angular Material Dialog](https://material.angular.dev/components/dialog)
- [NestJS Transactions](https://docs.nestjs.com/techniques/database#transactions)
- [Supabase RPC Functions](https://supabase.com/docs/guides/database/functions)
- [Zod Validation](https://zod.dev/)

## Implementation Notes
- Use Supabase RPC functions for atomic operations rather than manual transactions
- Follow project's French vocabulary: "prévisions" for budget lines
- Implement debouncing for rapid operations to prevent UI lag
- Use explicit signal updates after FormArray modifications
- Add comprehensive error handling with user-friendly messages

### PR DESCRIPTION
**Pull request by neogenz**
Close #55

### CODE REVIEW

File: frontend/projects/webapp/src/app/feature/budget-templates/details/components/edit-transactions-dialog.ts [hunk 1]
```diff
@@ -506,4 +445,56 @@ export default class EditTransactionsDialog {
...
+      data: {
+        title: 'Confirmer la suppression',
+        message: 'Êtes-vous sûr de vouloir supprimer cette ligne ?',
+        confirmText: 'Supprimer',
+        cancelText: 'Annuler',
+        confirmColor: 'warn',
+      } as ConfirmationDialogData,
+    });
+
+    return firstValueFrom(dialogRef.afterClosed()) || false;
```
Comment (neogenz) [inline frontend/projects/webapp/src/app/feature/budget-templates/details/components/edit-transactions-dialog.ts:491]: Cette fonction passe-plat est overkill

File: frontend/projects/webapp/src/app/feature/budget-templates/details/components/edit-transactions-dialog.ts [hunk 2]
```diff
@@ -506,4 +445,56 @@ export default class EditTransactionsDialog {
...
+      return;
+    }
+
+    this.#state.updateTransaction(transactionId, { [field]: value });
+  }
+
+  /**
+   * Show confirmation dialog for transaction removal
+   */
+  async #showConfirmationDialog(): Promise<boolean> {
```
Comment (neogenz) [inline frontend/projects/webapp/src/app/feature/budget-templates/details/components/edit-transactions-dialog.ts:451]: async sans await ?
